### PR TITLE
Fastmode record wildcards

### PR DESCRIPTION
**Pull request by OlliV**

### CODE REVIEW

File: server/modules/selva/module/selva_object.c [hunk 1]
```diff
@@ -1320,6 +1320,81 @@ int SelvaObject_ReplyWithObject(RedisModuleCtx *ctx, struct SelvaObject *obj, co
...
 
+int SelvaObject_GetWithWildcard(RedisModuleCtx *ctx, struct SelvaObject *obj, const char *okey_str, size_t okey_len, long *resp_count, int resp_path_start_idx, unsigned int flags) {
+    char *wildcard = strstr(okey_str, ".*.");
+    size_t wildcard_idx = wildcard - okey_str;
+    size_t idx = wildcard_idx + 1; // .*. => *.
+
+    // path before the wildcard character
+    const size_t before_len = idx - 1;
+    char before[before_len];
+    memcpy(before, &okey_str[0], before_len);
```
Comment (OlliV) [inline server/modules/selva/module/selva_object.c:1331]:
```suggestion
    const char *before = okey_str;
```

File: server/modules/selva/module/selva_object.c [hunk 2]
```diff
@@ -1320,6 +1320,81 @@ int SelvaObject_ReplyWithObject(RedisModuleCtx *ctx, struct SelvaObject *obj, co
...
+                char reply_path[reply_path_len];
+                if (resp_path_start_idx == -1) {
+                    sprintf(reply_path, "%.*s.%.*s", (int)obj_key_len, obj_key_name_str, (int)key->name_len, key->name);
+                } else {
+                    sprintf(reply_path, "%.*s.%.*s.%.*s", (int)before_len - resp_path_start_idx, before + resp_path_start_idx, (int)obj_key_len, obj_key_name_str, (int)key->name_len, key->name);
+                }
+                RedisModule_ReplyWithStringBuffer(ctx, reply_path, reply_path_len);
+            } else {
+                // if the whole resolved path should be returned
+                size_t reply_path_len = before_len + 1 + obj_key_len + 1 + key->name_len;
```
Comment (OlliV) [inline server/modules/selva/module/selva_object.c:1377]:
```suggestion
                const size_t reply_path_len = before_len + 1 + obj_key_len + 1 + key->name_len;
```

File: server/modules/selva/module/selva_object.c [hunk 3]
```diff
@@ -1320,6 +1320,81 @@ int SelvaObject_ReplyWithObject(RedisModuleCtx *ctx, struct SelvaObject *obj, co
...
+                size_t reply_path_len = resp_path_start_idx == -1 ? obj_key_len + 1 + key->name_len : (before_len - resp_path_start_idx) + 1 + obj_key_len + 1 + key->name_len;
+                char reply_path[reply_path_len];
+                if (resp_path_start_idx == -1) {
+                    sprintf(reply_path, "%.*s.%.*s", (int)obj_key_len, obj_key_name_str, (int)key->name_len, key->name);
+                } else {
+                    sprintf(reply_path, "%.*s.%.*s.%.*s", (int)before_len - resp_path_start_idx, before + resp_path_start_idx, (int)obj_key_len, obj_key_name_str, (int)key->name_len, key->name);
+                }
+                RedisModule_ReplyWithStringBuffer(ctx, reply_path, reply_path_len);
+            } else {
+                // if the whole resolved path should be returned
```
Comment (OlliV) [inline server/modules/selva/module/selva_object.c:1376]:
```suggestion
                /* If the whole resolved path should be returned. */
```

File: server/modules/selva/module/selva_object.c [hunk 4]
```diff
@@ -1320,6 +1320,81 @@ int SelvaObject_ReplyWithObject(RedisModuleCtx *ctx, struct SelvaObject *obj, co
...
+
+    // path after the wildcard character
+    const size_t after_len = okey_len - idx - 2;
+    char after[after_len];
+    memcpy(after, &okey_str[idx + 2], after_len);
+
+    struct SelvaObjectKey *key;
+    int err = 0;
+
+    err = get_key(obj, before, before_len, 0, &key);
```
Comment (OlliV) [inline server/modules/selva/module/selva_object.c:1341]: Could be also:

```
if (err) {
    return err;
}

❓

File: server/modules/selva/module/selva_object.c [hunk 5]
```diff
@@ -1320,6 +1320,81 @@ int SelvaObject_ReplyWithObject(RedisModuleCtx *ctx, struct SelvaObject *obj, co
...
+
+    // path before the wildcard character
+    const size_t before_len = idx - 1;
+    char before[before_len];
+    memcpy(before, &okey_str[0], before_len);
+
+    // path after the wildcard character
+    const size_t after_len = okey_len - idx - 2;
+    char after[after_len];
+    memcpy(after, &okey_str[idx + 2], after_len);
```
Comment (OlliV) [inline server/modules/selva/module/selva_object.c:1336]:
```suggestion
    const size_t after_len = okey_len - idx - 2;
    const char *after = &okey_str[idx + 2];
```

**Review by OlliV — COMMENTED** [review]
🚧

File: server/modules/selva/module/selva_object.c [hunk 6]
```diff
@@ -1320,6 +1320,75 @@ int SelvaObject_ReplyWithObject(RedisModuleCtx *ctx, struct SelvaObject *obj, co
...
+        const char *obj_key_name_str;
+        while ((obj_key_name_str = SelvaObject_ForeachKey(key->value, &it))) {
+
+            const size_t obj_key_len = strlen(obj_key_name_str);
+            size_t new_field_len = before_len + 1 + obj_key_len + 1 + after_len;
+            char new_field[new_field_len];
+            sprintf(new_field, "%.*s.%.*s.%.*s", (int)before_len, before, (int)obj_key_len, obj_key_name_str, (int)after_len, after);
+
+            if (strstr(new_field, ".*.")) {
+                SelvaObject_GetWithWildcard(ctx, obj, new_field, new_field_len, resp_count, resp_path_start_idx == -1 ? idx : resp_path_start_idx, flags);
```
[the remark this answers is not in the document]
Comment (jimdebeer) [reply server/modules/selva/module/selva_object.c]: indeed

**Pull request merged**